File: core/pkg/monitor/tpu.go
//go:build linux

package monitor

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/wandb/wandb/core/pkg/monitor/tpuproto"
	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/local"
)

// TPUMetricName represents a TPU metric name for querying on the gRPC server exposed by the TPU runtime.
type TPUMetricName string

const (
	// googleTPUVendorID is the PCI vendor ID assigned to Google TPUs.
	googleTPUVendorID = "0x1ae0"

	// grpcAddr is the gRPC server address for the TPU runtime.
	//
	// See https://github.com/google/cloud-accelerator-diagnostics/tree/main/tpu_info for more details.
	grpcAddr = "localhost:8431"

	// TPUTotalMemory is the total High Bandwidth Memory in bytes.
	TPUTotalMemory TPUMetricName = "tpu.runtime.hbm.memory.total.bytes"
	// TPUMemoryUsage is the current High Bandwidth Memory usage in bytes.
	TPUMemoryUsage TPUMetricName = "tpu.runtime.hbm.memory.usage.bytes"
	// TPUDutyCyclePct is the TensorCore duty cycle percentage.
	TPUDutyCyclePct TPUMetricName = "tpu.runtime.tensorcore.dutycycle.percent"
)

// TPUChip represents TPU chip specifications.
type TPUChip struct {
	Name           string // The name of the TPU chip (e.g., "v2", "v3").
	HbmGiB         int    // High Bandwidth Memory in GiB
	DevicesPerChip int    // Number of devices per chip
}

type RuntimeMetricServiceClient interface {
	GetRuntimeMetric(ctx context.Context, in *tpuproto.MetricRequest, opts ...grpc.CallOption) (*tpuproto.MetricResponse, error)
}

// TPU represents a TPU asset with gRPC connection and client.
//
// This code is based on Google's Cloud Accelerator Diagnostics project:
// https://github.com/google/cloud-accelerator-diagnostics.
type TPU struct {
	// Name of the TPU asset used to identify the asset in the metadata.
	name string

	// gRPC client connection and client for runtime metrics.
	//
	// TPU runtime metrics are exposed via a gRPC server running on a Google Cloud TPU VM.
	conn   *grpc.ClientConn
	client RuntimeMetricServiceClient

	// TPU chip specifications.
	chip *TPUChip

	// Total number of TPU devices detected.
	count int
}

// NewTPU creates a new TPU instance by detecting local TPU chips and initializing the gRPC connection.
func NewTPU() *TPU {
	t := &TPU{name: "tpu"}

	chip, count := getLocalTPUChips()
	if chip == nil {
		return nil
	}
	t.chip = chip
	t.count = count

	conn, err := grpc.NewClient(grpcAddr, grpc.WithTransportCredentials(local.NewCredentials()))
	if err != nil {
		return nil
	}
	client := tpuproto.NewRuntimeMetricServiceClient(conn)
	t.conn = conn
	t.client = client

	return t
}

func (t *TPU) Name() string {
	return t.name
}

func (t *TPU) SetName(name string) {
	t.name = name
}

func (t *TPU) SetChip(chip *TPUChip, count int) {
	t.chip = chip
	t.count = count
}

func (t *TPU) SetClient(client RuntimeMetricServiceClient) {
	t.client = client
}

// Sample returns TPU metrics such as memory usage in % and in bytes, and duty cycle.
func (t *TPU) Sample() (map[string]any, error) {
	if t.client == nil || t.chip == nil {
		return nil, nil
	}

	// Total memory per TPU core [bytes]
	totals, err := t.getMetrics(TPUTotalMemory)
	if err != nil {
		return nil, err
	}
	// Memory usage per TPU core [bytes]
	usages, err := t.getMetrics(TPUMemoryUsage)
	if err != nil {
		return nil, err
	}
	// Duty cycle per TPU device [%]
	dutyCycles, err := t.getMetrics(TPUDutyCyclePct)
	if err != nil {
		return nil, err
	}

	// See below for the expected number of metrics per chip
	if len(totals) != len(usages) || len(usages) != len(dutyCycles)*t.chip.DevicesPerChip {
		return nil, fmt.Errorf("tpu: metrics not found for all chips")
	}

	memoryUsages := make(map[int64]int64)
	for _, usage := range usages {
		deviceID := usage.GetAttribute().GetValue().GetIntAttr()
		memoryUsage := usage.GetGauge().GetAsInt()
		memoryUsages[deviceID] = memoryUsage
	}

	totalMemories := make(map[int64]int64)
	for _, total := range totals {
		deviceID := total.GetAttribute().GetValue().GetIntAttr()
		totalMemory := total.GetGauge().GetAsInt()
		totalMemories[deviceID] = totalMemory
	}

	// Duty cycle is measured per chip; distribute it to each device (core) in the chip.
	dutyCyclesPerCore := make(map[int64]float64)
	for _, duty := range dutyCycles {
		chipID := duty.GetAttribute().GetValue().GetIntAttr()
		dutyCycle := duty.GetGauge().GetAsDouble()
		dutyCyclesPerCore[chipID*int64(t.chip.DevicesPerChip)] = dutyCycle
		dutyCyclesPerCore[chipID*int64(t.chip.DevicesPerChip)+1] = dutyCycle
	}

	data := make(map[string]any)

	for deviceID := int64(0); deviceID < int64(t.count); deviceID++ {
		memoryUsage, ok := memoryUsages[deviceID]
		if !ok {
			continue
		}

		totalMemory, ok := totalMemories[deviceID]
		if !ok {
			continue
		}

		dutyCycle, ok := dutyCyclesPerCore[deviceID]
		if !ok {
			continue
		}

		// Memory usage [%]
		memoryUsageKey := fmt.Sprintf("%s.%d.memoryUsage", t.Name(), deviceID)
		// Memory usage [bytes]
		memoryUsageBytesKey := fmt.Sprintf("%s.%d.memoryUsageBytes", t.Name(), deviceID)
		// Duty cycle [%]
		dutyCycleKey := fmt.Sprintf("%s.%d.dutyCycle", t.Name(), deviceID)

		data[memoryUsageKey] = float64(memoryUsage) / float64(totalMemory) * 100
		data[memoryUsageBytesKey] = memoryUsage
		data[dutyCycleKey] = dutyCycle
	}

	return data, nil
}

func (t *TPU) IsAvailable() bool {
	return t.chip != nil
}

// Close closes the gRPC connection and releases resources.
func (t *TPU) Close() {
	if t.conn != nil {
		t.conn.Close()
		t.conn = nil
		t.client = nil
	}
}

// getLocalTPUChips scans the PCI devices to detect local TPU chips and
// returns the most common chip type and the total count.
func getLocalTPUChips() (*TPUChip, int) {
	devices, err := filepath.Glob("/sys/bus/pci/devices/*")
	if err != nil {
		return nil, 0
	}

	counter := make(map[*TPUChip]int)

	for _, pciPath := range devices {
		vendorPath := filepath.Join(pciPath, "vendor")
		data, err := os.ReadFile(vendorPath)
		if err != nil {
			continue
		}
		vendorID := strings.TrimSpace(string(data))
		if vendorID != googleTPUVendorID {
			continue
		}

		devicePath := filepath.Join(pciPath, "device")
		data, err = os.ReadFile(devicePath)
		if err != nil {
			continue
		}
		deviceID := strings.TrimSpace(string(data))

		subsystemPath := filepath.Join(pciPath, "subsystem_device")
		data, err = os.ReadFile(subsystemPath)
		if err != nil {
			continue
		}
		subsystemID := strings.TrimSpace(string(data))

		chipType, err := tpuChipFromPCIDeviceID(deviceID, subsystemID)
		if err != nil {
			continue
		}

		counter[chipType]++
	}

	if len(counter) == 0 {
		return nil, 0
	}

	var mostCommonChip *TPUChip
	var maxCount int
	for chip, count := range counter {
		if count > maxCount {
			mostCommonChip = chip
			maxCount = count
		}
	}
	return mostCommonChip, maxCount
}

func tpuChipFromPCIDeviceID(deviceID, subsystemID string) (*TPUChip, error) {
	switch deviceID {
	case "0x0027":
		switch subsystemID {
		case "0x004e":
			return &TPUChip{Name: "v2", HbmGiB: 8, DevicesPerChip: 2}, nil
		case "0x004f":
			return &TPUChip{Name: "v3", HbmGiB: 16, DevicesPerChip: 2}, nil
		}
	case "0x005e":
		return &TPUChip{Name: "v4", HbmGiB: 32, DevicesPerChip: 1}, nil
	case "0x0063":
		return &TPUChip{Name: "v5e", HbmGiB: 16, DevicesPerChip: 1}, nil
	case "0x0062":
		return &TPUChip{Name: "v5p", HbmGiB: 95, DevicesPerChip: 1}, nil
	case "0x006f":
		return &TPUChip{Name: "v6e", HbmGiB: 32, DevicesPerChip: 1}, nil
	}

	return nil, fmt.Errorf("unknown TPU chip")
}

// getMetrics retrieves metrics from the TPU runtime gRPC service for the given metric name.
func (t *TPU) getMetrics(metricName TPUMetricName) ([]*tpuproto.Metric, error) {
	req := &tpuproto.MetricRequest{MetricName: string(metricName)}

	resp, err := t.client.GetRuntimeMetric(context.Background(), req)
	if err != nil {
		return nil, err
	}
	metrics := resp.Metric.Metrics

	return metrics, nil
}

// Probe returns the TPU metadata.
func (t *TPU) Probe() *spb.MetadataRequest {
	if t.chip == nil {
		return nil
	}

	return &spb.MetadataRequest{
		Tpu: &spb.TPUInfo{
			Name:           t.chip.Name,
			Count:          uint32(t.count),
			HbmGib:         uint32(t.chip.HbmGiB),
			DevicesPerChip: uint32(t.chip.DevicesPerChip),
		},
	}
}

File: gpu_stats/src/tpu.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::sync::Arc;
use tonic::transport::{Channel, ClientTlsConfig};
use tonic::Request;

mod tpuproto {
    tonic::include_proto!("tpuproto");
}

use tpuproto::{runtime_metric_service_client::RuntimeMetricServiceClient, MetricRequest};

const GOOGLE_TPU_VENDOR_ID: &str = "0x1ae0";
const GRPC_ADDR: &str = "http://localhost:8431";

#[derive(Debug, Clone)]
struct TPUChip {
    name: String,
    hbm_gib: u32,
    devices_per_chip: u32,
}

#[derive(Debug)]
struct TPU {
    name: String,
    client: Option<RuntimeMetricServiceClient<Channel>>,
    chip: TPUChip,
    count: usize,
}

impl TPU {
    async fn new() -> Option<Self> {
        let (chip, count) = Self::get_local_tpu_chips();
        if count == 0 {
            return None;
        }

        let channel = Channel::from_static(GRPC_ADDR).connect().await.ok()?;
        let client = RuntimeMetricServiceClient::new(channel);

        Some(TPU {
            name: "tpu".to_string(),
            client: Some(client),
            chip,
            count,
        })
    }

    async fn sample(&mut self) -> Result<HashMap<String, f64>, Box<dyn std::error::Error>> {
        let client = self
            .client
            .as_mut()
            .ok_or("TPU client is not initialized")?;

        let total_memory = self
            .get_metrics(client, "tpu.runtime.hbm.memory.total.bytes")
            .await?;
        let memory_usage = self
            .get_metrics(client, "tpu.runtime.hbm.memory.usage.bytes")
            .await?;
        let duty_cycle = self
            .get_metrics(client, "tpu.runtime.tensorcore.dutycycle.percent")
            .await?;

        let mut metrics = HashMap::new();
        for (&device_id, &usage) in &memory_usage {
            let total = total_memory.get(&device_id).copied().unwrap_or(1.0);
            metrics.insert(
                format!("{}.{}.memoryUsage", self.name, device_id),
                usage / total * 100.0,
            );
            metrics.insert(
                format!("{}.{}.memoryUsageBytes", self.name, device_id),
                usage,
            );
        }
        for (&device_id, &duty) in &duty_cycle {
            metrics.insert(format!("{}.{}.dutyCycle", self.name, device_id), duty);
        }

        Ok(metrics)
    }

    async fn get_metrics(
        &self,
        client: &mut RuntimeMetricServiceClient<Channel>,
        metric_name: &str,
    ) -> Result<HashMap<i64, f64>, Box<dyn std::error::Error>> {
        let request = Request::new(MetricRequest {
            metric_name: metric_name.to_string(),
        });
        let response = client.get_runtime_metric(request).await?.into_inner();

        let mut metrics = HashMap::new();
        for metric in response.metric.unwrap().metrics {
            let device_id = metric.attribute.unwrap().value.unwrap().int_attr;
            let value = metric.gauge.unwrap().as_double.unwrap();
            metrics.insert(device_id, value);
        }
        Ok(metrics)
    }

    fn get_local_tpu_chips() -> (TPUChip, usize) {
        let pci_devices = fs::read_dir("/sys/bus/pci/devices").ok()?;
        let mut counter = HashMap::new();

        for entry in pci_devices.flatten() {
            let path = entry.path();
            let vendor_id = fs::read_to_string(path.join("vendor"))
                .ok()?
                .trim()
                .to_string();
            if vendor_id != GOOGLE_TPU_VENDOR_ID {
                continue;
            }
            let device_id = fs::read_to_string(path.join("device"))
                .ok()?
                .trim()
                .to_string();
            let subsystem_id = fs::read_to_string(path.join("subsystem_device"))
                .ok()?
                .trim()
                .to_string();

            if let Ok(chip) = Self::tpu_chip_from_pci_device_id(&device_id, &subsystem_id) {
                *counter.entry(chip).or_insert(0) += 1;
            }
        }

        counter
            .into_iter()
            .max_by_key(|&(_, count)| count)
            .unwrap_or_default()
    }

    fn tpu_chip_from_pci_device_id(
        device_id: &str,
        subsystem_id: &str,
    ) -> Result<TPUChip, &'static str> {
        match (device_id, subsystem_id) {
            ("0x0027", "0x004e") => Ok(TPUChip {
                name: "v2".to_string(),
                hbm_gib: 8,
                devices_per_chip: 2,
            }),
            ("0x0027", "0x004f") => Ok(TPUChip {
                name: "v3".to_string(),
                hbm_gib: 16,
                devices_per_chip: 2,
            }),
            ("0x005e", _) => Ok(TPUChip {
                name: "v4".to_string(),
                hbm_gib: 32,
                devices_per_chip: 1,
            }),
            ("0x0063", _) => Ok(TPUChip {
                name: "v5e".to_string(),
                hbm_gib: 16,
                devices_per_chip: 1,
            }),
            ("0x0062", _) => Ok(TPUChip {
                name: "v5p".to_string(),
                hbm_gib: 95,
                devices_per_chip: 1,
            }),
            ("0x006f", _) => Ok(TPUChip {
                name: "v6e".to_string(),
                hbm_gib: 32,
                devices_per_chip: 1,
            }),
            _ => Err("Unknown TPU chip"),
        }
    }
}
